Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1333 @@
+          ↙   ↘
+        1       +
+              ↙   ↘
+            2       3
+```
+
+Which was generated from this grammar:
+
+```
+exp -> num
+exp -> num "+" exp
+num -> [0-9]+
+```
+
+It means that in an expression such as `1 + 2 + 3`, a LL parser would perform substitutions by visiting the leftmost nodes first, recovering the tokens in this order:
+
+<pre>
+<i>exp</i>
+<b><i>num</i> + exp</b>
+<b>1</b> + <i>exp</i>
+1 + <b><i>num</i> + exp</b>
+1 + <b>2</b> + <i>exp</i>
+1 + 2 + <b><i>num</i></b>
+1 + 2 + <b>3</b>
+</pre>
+
+n.b.: **Bold** means the result of the derivation, and <i>italic</i> means the derivation that will be expanded next.
+
+Whilst an LR parser would do the following:
+
+<pre>
+<i>exp</i>
+<b>num + <i>exp</i></b>
+num + <b>num + <i>exp</i></b>
+num + num + <b><i>num</i></b>
+num + <i>num</i> + <b>3</b>
+<i>num</i> + <b>2</b> + 3
+<b>1</b> + 2 + 3
+</pre>
+
+For more information on leftmost and rightmost derivations, see [LL and LR Parsing Demystified](https://blog.reverberate.org/2013/07/ll-and-lr-parsing-demystified.html).
+
+Another difference is that Alex is a lexer generator and Happy a parser generator.
+It means that instead of writing Haskell files (`.hs`) like in parser combinators, you will write Alex (`.x`) and Happy (`.y`) files, which in turn will automatically generate Haskell files with implementations of the lexer and parser recognizing the described grammar.
+
+Below are some pros and cons of Happy in comparison to, for example, Megaparsec:
+
+Pros:
+* LR parsers are more powerful than LL, allowing to parse more complex grammars with greater ease.
+* Ambiguous grammars are reported to the user, allowing a greater trust in the parser's output and saving the debugging headache.
+* LR parsers can handle left-recursion, while LL parsers can't.
+
+Cons:
+* Less idiomatic, as you need to write Alex and Happy grammars instead of Haskell code.
+* Less flexible since you can't touch the algorithm on how Happy works. Meanwhile, you can easily create your functions for parser combinators.
+
+Comparing the implementations of both parsing techniques is beyond the scope of this post. Still, if you are interested, you may read the parser combinators article linked above.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial, but for this, we recommend you read our article, [How to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminals, non-terminals, and production rules
+
+A **terminal** is a lexical item of the language; in our case they will be tokens that were produced by Alex.
+
+A **non-terminal** is an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations.
+In the example above, `myNonTerminal` is non-terminal.
```

Comment:
"above" should be changed to "below"?

Reply:
Good catch, thanks for pointing it out. :+1: 

By the way, just curious, but how did you find this repo? We're happy (pun intended) because you're our first external reviewer in this repo.

Reply:
Ah, I see, thanks for explaining. This is a pretty small world (the world of Haskell, that is), at least I feel like we frequently find the same people in different places.

This gives me an idea, @NaeosPsy, what do you think if we start linking to our GitHub repos? At the bottom of the articles, we write something like "Found a typo or something incorrect? Feel free to open an issue in our GitHub repo."

And thanks for mentioning the typo, we'll fix it.